services/metadata_service/api/run.py

```diff
@@ -136,6 +143,97 @@ async def create_run(self, request):
 
         return await self._async_table.add_run(run_row, fill_heartbeat=client_supports_heartbeats)
 
+    @format_response
+    @handle_exceptions
+    async def mutate_user_tags(self, request):
+        """
+        ---
+        description: mutate user tags
+        tags:
+        - Run
+        parameters:
+        - name: "flow_id"
+          in: "path"
+          description: "flow_id"
+          required: true
+          type: "string"
+        - name: "run_number"
+          in: "path"
+          description: "run_number"
+          required: true
+          type: "string"
+        - name: "body"
+          in: "body"
+          description: "body"
+          required: true
+          schema:
+            type: object
+            properties:
+                tags_to_add:
+                    type: array of string
+                tags_to_remove:
+                    type: array of string
+        produces:
+        - 'text/plain'
+        responses:
+            "200":
+                description: successful operation. Tags updated.  Returns latest user tags
+            "400":
+                description: invalid HTTP Request
+            "405":
+                description: invalid HTTP Method
+        """
+        flow_name = request.match_info.get("flow_id")
+        run_number = request.match_info.get("run_number")
+        body = await read_body(request.content)
+        tags_to_add = body.get("tags_to_add", [])
+        tags_to_remove = body.get("tags_to_remove", [])
+
+        if not isinstance(tags_to_add, list):
+            return DBResponse(response_code=422, body="tags_to_add must be a list")
+
+        if not isinstance(tags_to_remove, list):
+            return DBResponse(response_code=422, body="tags_to_remove must be a list")
+
+        # let's make sure we have a list of strings
+        if not all(isinstance(t, str) for t in chain(tags_to_add, tags_to_remove)):
+            return DBResponse(response_code=422, body="All tag values must be strings")
+
+        tags_to_add_set = set(tags_to_add)
+        tags_to_remove_set = set(tags_to_remove)
+
+        async def _in_tx_mutation_logic(cur):
+            run_db_response = await self._async_table.get_run(flow_name, run_number, cur=cur)
+            if run_db_response.response_code != 200:
+                if run_db_response.response_code == 404:
+                    return run_db_response
+                return DBResponse(response_code=422,
+                                  body="Failed to get run (get_run status %d)" % run_db_response.response_code)
+            run = run_db_response.body
+            existing_tag_set = set(run["tags"])
+            existing_system_tag_set = set(run["system_tags"])
+
+            for tag in tags_to_remove_set:
```

Comment:
same as for local modificationn; this can be a simpler set operation and return all errors at once instead of one at a time.

Reply:
Done!